src/buffer/out/textBuffer.cpp

```diff
@@ -2235,7 +2272,54 @@ SHORT TextBuffer::GetCurHyperlinkId() const noexcept
 // - The hyperlink ID
 // Return Value:
 // - The URI
-std::wstring TextBuffer::GetHyperlinkUriFromId(SHORT id) const noexcept
+std::wstring TextBuffer::GetHyperlinkUriFromId(USHORT id) const
 {
     return _hyperlinkMap.at(id);
 }
+
+// Method description:
+// - Provides the hyperlink ID to be assigned as a text attribute, based on the optional custom id provided
+// Arguments:
+// - The user-defined id
+// Return value:
+// - The internal hyperlink ID
+USHORT TextBuffer::GetHyperlinkId(std::wstring_view params)
+{
+    USHORT id = 0;
+    if (params.empty())
+    {
+        // no custom id specified, return our internal count
+        id = _currentHyperlinkId;
+        ++_currentHyperlinkId;
+    }
+    else
+    {
+        // assign _currentHyperlinkId if the custom id does not already exist
+        const auto result = _customIdMap.emplace(params, _currentHyperlinkId);
+        if (result.second)
+        {
+            // the custom id did not already exist
+            ++_currentHyperlinkId;
+        }
+        id = (*(result.first)).second;
+    }
+    return id;
+}
+
+// Method Description:
+// - Removes a hyperlink from the hyperlink map and the associated
+//   user defined id from the custom id map (if there is one)
+// Arguments:
+// - The ID of the hyperlink to be removed
+void TextBuffer::RemoveHyperlinkFromMap(USHORT id)
+{
+    _hyperlinkMap.erase(id);
+    for (auto it = _customIdMap.begin(); it != _customIdMap.end(); ++it)
+    {
+        if (it->second == id)
+        {
+            _customIdMap.erase(it->first);

```

Comment:
For anyone else who had the same thought I did here, apparently the iterators and ordering is stable when iterating/erasing a `std::unordered_map` unlike some other STL structures: https://en.cppreference.com/w/cpp/container/unordered_map/erase

(This is a no-op for you, Pankaj. This has just come up before with trying to erase things inside the same loop as an iterator before.)